### PR TITLE
Experimental: type level non empty arrays

### PR DESCRIPTION
**Pull request by gcanti**
@yawaramin @sledorze here's a candidate implementation, could you please review?

### CODE REVIEW

File: src/NonEmptyArray2v.ts [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+export function tail<A>(nea: NonEmptyArray<A>): Array<A> {
+  return nea.slice(1)
+}
+
+/**
+ * @since 1.14.4
+ */
+export function min<A>(ord: Ord<A>): (nea: NonEmptyArray<A>) => A {
+  const foldO = fold(getMeetSemigroup(ord))
+  return nea => foldO(nea[0])(nea.slice(1))
```
[the remark this answers is not in the document]
Comment (gcanti) [reply src/NonEmptyArray2v.ts]: What about

```ts
export function min<A>(ord: Ord<A>): (nea: NonEmptyArray<A>) => A {
  const S = getMeetSemigroup(ord)
  return nea => nea.reduce(S.concat)
}
```

instead?

File: src/NonEmptyArray2v.ts [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+ */
+export function sort<A>(O: Ord<A>): (nea: NonEmptyArray<A>) => NonEmptyArray<A> {
+  return A.sort(O) as any
+}
+
+/**
+ * @since 1.14.4
+ */
+export function lookup<A>(i: number, nea: NonEmptyArray<A>): Option<A> {
+  return i === 0 ? some(nea[0]) : A.lookup(i - 1, nea.slice(1))
```
[the remark this answers is not in the document]
Comment (gcanti) [reply src/NonEmptyArray2v.ts]: Thinking twice, we can actually drop `lookup`, it's the same as `A.lookup`

File: src/NonEmptyArray2v.ts [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    return r
+  }
+  let head: A = as[0]
+  let tail: Array<A> = []
+  for (let i = 1; i < len; i++) {
+    const x = as[i]
+    if (S.equals(x, head)) {
+      tail.push(x)
+    } else {
+      r.push(make(head, tail))
```
[the remark this answers is not in the document]
Comment (gcanti) [reply src/NonEmptyArray2v.ts]: Also we can move `const r: Array<NonEmptyArray<A>>` after the first `if` and leverage `fromNonEmptyArray`

```ts
export const group = <A>(S: Setoid<A>) => (as: Array<A>): Array<NonEmptyArray<A>> => {
  const len = as.length
  if (len === 0) {
    return A.empty
  }
  const r: Array<NonEmptyArray<A>> = []
  let head: A = as[0]
  let newNE = fromNonEmptyArray([head])
  for (let i = 1; i < len; i++) {
    const x = as[i]
    if (S.equals(x, head)) {
      newNE.push(x)
    } else {
      r.push(newNE)
      head = x
      newNE = fromNonEmptyArray([head])
    }
  }
  r.push(newNE)
  return r
}
```

File: src/NonEmptyArray2v.ts [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+  map: <B>(f: (a: A, index: number, nea: NonEmptyArray<A>) => B) => NonEmptyArray<B>
+  concat: (as: Array<A>) => NonEmptyArray<A>
+  reverse: () => NonEmptyArray<A>
+}
+
+/**
+ * @since 1.14.5
+ */
+export function make<A>(head: A, tail: Array<A>): NonEmptyArray<A> {
+  return [head, ...tail] as any
```
Comment (yawaramin) [inline src/NonEmptyArray2v.ts]: I believe this can be `return <NonEmptyArray<A>>[head, ...tail]`.

Comment (gcanti) [reply src/NonEmptyArray2v.ts]: @yawaramin thanks for reviewing

```
<NonEmptyArray<A>>[head, ...tail]
```

raises the following error

```
Property '0' is missing in type 'A[]' but required in type 'NonEmptyArray<A>'
```

I'm afraid that casting to `any` is the only way to please TypeScript here, is not a big deal though

**Comment by gcanti** [comment]
@yawaramin @sledorze I'm going to merge this, my plan is:

- do **not** deprecate `NonEmptyArray` for now
- release `NonEmptyArray2v` in 1.15 as experimental feature so we try it out in the wild

**Pull request merged**